tests/run-make/output-type-permutations/rmake.rs

```diff
@@ -0,0 +1,177 @@
+// In 2014, rustc's output flags were reworked to be a lot more modular.
+// This test uses these output flags in an expansive variety of combinations
+// and syntax styles, checking that compilation is successful and that no unexpected
+// files are created.
+// The assert_eq! checks that "1 file remains" at the end of each part of the test,
+// because foo.rs counts as a file, and should be the only remaining one.
+// See https://github.com/rust-lang/rust/pull/12020
+
+use run_make_support::{
+    bin_name, cwd, dynamic_lib_name, fs_wrapper, rust_lib_name, rustc, static_lib_name,
+};
+
+fn remove_artifacts() {
+    std::fs::remove_file("libbar.ddl.exp").unwrap_or_default();
+    std::fs::remove_file("libbar.dll.lib").unwrap_or_default();
+    std::fs::remove_file("libbar.pdb").unwrap_or_default();
+    std::fs::remove_file("libbar.dll.a").unwrap_or_default();
+    std::fs::remove_file("libbar.exe.a").unwrap_or_default();
+    std::fs::remove_file("bar.ddl.exp").unwrap_or_default();
+    std::fs::remove_file("bar.dll.lib").unwrap_or_default();
+    std::fs::remove_file("bar.pdb").unwrap_or_default();
+    std::fs::remove_file("bar.dll.a").unwrap_or_default();
+    std::fs::remove_file("bar.exe.a").unwrap_or_default();
+}
+
+fn main() {
+    rustc().input("foo.rs").crate_type("rlib,dylib,staticlib").run();
+    fs_wrapper::remove_file(rust_lib_name("bar"));
+    fs_wrapper::remove_file(dynamic_lib_name("bar"));
+    fs_wrapper::remove_file(static_lib_name("bar"));
+    remove_artifacts();
+    assert_eq!(fs_wrapper::read_dir(cwd()).count(), 1);
```

Comment:
Suggestion: I want to say this check in itself is not sufficient, as my reading of the test comment we want to be checking two conditions here:

(1) unexpected files are not created
(2) expected files are created

Checks like `assert_eq!(fs_wrapper::read_dir(cwd()).count(), 1);` are only a *proxy* for checking for the correct *number* of files, but it does not actually check what the precise files are.

I would imagine the test looking something like

```rs
fn assert_expected_output_files(
    expected_output_files: &[&'static str],
    rustc_invocation: Fn())
{
    // or somehow set output directory
    run_in_tmpdir(|| {
        rustc_invocation();
        let actual_output_files = find_files_in_directory(); // pseudocode
        assert_eq!(&actual_output_files, expected_output_files, "output artifacts mismatch");
    });
}
```